libraries/chain/include/eosio/chain/block.hpp

```diff
@@ -103,6 +103,79 @@ namespace eosio { namespace chain {
    };
    using signed_block_ptr = std::shared_ptr<signed_block>;
 
+   struct pruned_transaction_receipt : public transaction_receipt_header {
+
+      pruned_transaction_receipt():transaction_receipt_header(){}
+      pruned_transaction_receipt(const transaction_receipt&, bool legacy);
+      explicit pruned_transaction_receipt( const transaction_id_type& tid ):transaction_receipt_header(executed),trx(tid){}
+      explicit pruned_transaction_receipt( const pruned_transaction& ptrx ):transaction_receipt_header(executed),trx(ptrx){}
+
+      fc::static_variant<transaction_id_type, pruned_transaction> trx;
+
+      std::size_t maximum_pruned_pack_size( pruned_transaction::compression_type segment_compression ) const;
+
+      digest_type digest()const {
+         digest_type::encoder enc;
+         fc::raw::pack( enc, status );
+         fc::raw::pack( enc, cpu_usage_us );
+         fc::raw::pack( enc, net_usage_words );
+         if( trx.contains<transaction_id_type>() )
+            fc::raw::pack( enc, trx.get<transaction_id_type>() );
+         else
+            fc::raw::pack( enc, trx.get<pruned_transaction>().packed_digest() );
+         return enc.result();
+      }
+   };
+
+   struct pruned_block : public signed_block_header{
+   private:
+      pruned_block( const pruned_block& ) = default;
+   public:
+      enum class prune_state_type : uint8_t { incomplete, complete, complete_legacy };
+
+      pruned_block() = default;
+      explicit pruned_block( const signed_block_header& h ):signed_block_header(h){}
+      pruned_block( const signed_block&, bool legacy );
+      pruned_block( pruned_block&& ) = default;
+      pruned_block& operator=(const pruned_block&) = delete;
+      pruned_block clone() const { return *this; }
+
+      fc::enum_type<uint8_t,prune_state_type> prune_state{prune_state_type::complete_legacy};
+      deque<pruned_transaction_receipt>       transactions; /// new or generated transactions
+      extensions_type                         block_extensions;
+
+      std::size_t maximum_pruned_pack_size( pruned_transaction::compression_type segment_compression ) const;
+
+      template<typename Stream>
+      void pack_with_padding(Stream& stream, pruned_transaction::compression_type segment_compression) {
+         std::size_t padded_size = maximum_pruned_pack_size( segment_compression );
+         pack_with_padding(stream, segment_compression, padded_size);
+      }
+      template<typename Stream>
+      void pack_with_padding(Stream& stream, pruned_transaction::compression_type segment_compression, std::size_t padded_size) {
+         fc::raw::pack(stream, fc::unsigned_int(padded_size));
```

Comment:
I specified a `uint32_t` rather than an `fc::unsigned_int` because I wanted to be able to serialize the `pruned_block` to disk in one pass (and a quick correction of the serialized `padded_size`). I suppose technically we could do the same with `fc::unsigned_int` (just reserve 5 bytes instead of 4 bytes) as long as we are not strict on the deserialization (which we currently aren't).

I think this two pass system for serialization will especially be problematic when you take into account compression. It would be terrible if we had to compress twice (the first time just to calculate the padding size and throw away all that effort).

Reply:
I missed that, because I was only looking at the spec for pruned_block.  Anyway, it's actually worse than that.  block_log.cpp currently packs the block into a vector before writing it to disk, since this version of fc::raw::pack calculates the size ahead of time, this will result in compressing the data up to 4 times (It could be reduced to 3 with an overload of pack_with_padding for datastream<size_t>).